### PR TITLE
Add support for config inheritance

### PR DESCRIPTION
**Pull request by apazzolini**
- Adds `inheritable` and `inherits` properties to Configs. 
- Additionally, adds `inheritedBy` for Configs when fetched individually.
- Allows toggling inheritability with `doppler configs update --inheritable [true/false]`
- Allows setting parent inheritances with `doppler configs update --inherits 'billing.dev,database.dev'`

Depends on https://github.com/DopplerHQ/server/pull/7068

Closes ENG-8534

### CODE REVIEW

File: pkg/cmd/configs.go [hunk 1]
```diff
@@ -196,34 +206,70 @@ func deleteConfigs(cmd *cobra.Command, args []string) {
 
 func updateConfigs(cmd *cobra.Command, args []string) {
 	jsonFlag := utils.OutputJSON
+
+	nameSet := cmd.Flags().Changed("name")
+	inheritableSet := cmd.Flags().Changed("inheritable")
+	inheritsSet := cmd.Flags().Changed("inherits")
+
+	if (nameSet && (inheritableSet || inheritsSet)) || (inheritableSet && inheritsSet) {
```
Comment (nmanoogian) [inline pkg/cmd/configs.go]: minor: This does not throw an error if no flags are set. Here's an alternative:

```diff
diff --git a/pkg/cmd/configs.go b/pkg/cmd/configs.go
index 3a4875f..1e3e3f1 100644
--- a/pkg/cmd/configs.go
+++ b/pkg/cmd/configs.go
@@ -211,7 +211,14 @@ func updateConfigs(cmd *cobra.Command, args []string) {
 	inheritableSet := cmd.Flags().Changed("inheritable")
 	inheritsSet := cmd.Flags().Changed("inherits")
 
-	if (nameSet && (inheritableSet || inheritsSet)) || (inheritableSet && inheritsSet) {
+	varsChanged := 0
+	for _, v := range []bool{nameSet, inheritableSet, inheritsSet} {
+		if v {
+			varsChanged++
+		}
+	}
+
+	if varsChanged != 1 {
 		utils.HandleError(fmt.Errorf("Exactly one of name, inheritable, and inherits must be specified"))
 	}
 


```

File: pkg/cmd/configs.go [hunk 2]
```diff
@@ -61,9 +61,19 @@ var configsDeleteCmd = &cobra.Command{
...
 	Short:             "Update a config",
+	Long:              "Update properties about a config, such as its name, inheritability flag, and inheritances",
 	Args:              cobra.MaximumNArgs(1),
 	ValidArgsFunction: configNamesValidArgs,
 	Run:               updateConfigs,
+	Example: `Updating a config's name
+$ doppler configs update --project proj --config dev_branch --name dev_branch2
+
+Enabling a config to be inherited
+$ doppler configs update --project proj --config dev --inheritable true
```
Comment (nmanoogian) [inline pkg/cmd/configs.go]: minor: Boolean flags behave a bit differently with this parsing lib:

```
# `true` and `false` are treated as separate positional arguments
$ doppler configs -p weather-service update -c dev --inheritable true
Unable to update config
Doppler Error: Could not find requested config 'true'
$ doppler configs -p weather-service update -c dev --inheritable false
Unable to update config
Doppler Error: Could not find requested config 'false'

# `--inheritable` on its own is treated as a true
$ doppler configs -p weather-service update -c dev --inheritable
┌──────┬───────────────┬────────────┬──────────────────────────┬─────────────┬─────────────────┬─────────────┬──────────────┐
│ NAME │ INITIAL FETCH │ LAST FETCH │ CREATED AT               │ ENVIRONMENT │ PROJECT         │ INHERITABLE │ INHERITED BY │
├──────┼───────────────┼────────────┼──────────────────────────┼─────────────┼─────────────────┼─────────────┼──────────────┤
│ dev  │               │            │ 2024-12-04T22:25:09.031Z │ dev         │ weather-service │ TRUE        │ <NONE>       │
└──────┴───────────────┴────────────┴──────────────────────────┴─────────────┴─────────────────┴─────────────┴──────────────┘

# `true` and `false` can be provided with `--inheritable=<boolean>`
$ doppler configs -p weather-service update -c dev --inheritable=true
┌──────┬───────────────┬────────────┬──────────────────────────┬─────────────┬─────────────────┬─────────────┬──────────────┐
│ NAME │ INITIAL FETCH │ LAST FETCH │ CREATED AT               │ ENVIRONMENT │ PROJECT         │ INHERITABLE │ INHERITED BY │
├──────┼───────────────┼────────────┼──────────────────────────┼─────────────┼─────────────────┼─────────────┼──────────────┤
│ dev  │               │            │ 2024-12-04T22:25:09.031Z │ dev         │ weather-service │ TRUE        │ <NONE>       │
└──────┴───────────────┴────────────┴──────────────────────────┴─────────────┴─────────────────┴─────────────┴──────────────┘
$ doppler configs -p weather-service update -c dev --inheritable=false
┌──────┬───────────────┬────────────┬──────────────────────────┬─────────────┬─────────────────┬─────────────┬──────────┐
│ NAME │ INITIAL FETCH │ LAST FETCH │ CREATED AT               │ ENVIRONMENT │ PROJECT         │ INHERITABLE │ INHERITS │
├──────┼───────────────┼────────────┼──────────────────────────┼─────────────┼─────────────────┼─────────────┼──────────┤
│ dev  │               │            │ 2024-12-04T22:25:09.031Z │ dev         │ weather-service │ FALSE       │ <NONE>   │
└──────┴───────────────┴────────────┴──────────────────────────┴─────────────┴─────────────────┴─────────────┴──────────┘
```

For the purposes of this example, I might recommend:

```diff
diff --git a/pkg/cmd/configs.go b/pkg/cmd/configs.go
index 3a4875f..e455607 100644
--- a/pkg/cmd/configs.go
+++ b/pkg/cmd/configs.go
@@ -69,7 +69,7 @@ var configsUpdateCmd = &cobra.Command{
 $ doppler configs update --project proj --config dev_branch --name dev_branch2
 
 Enabling a config to be inherited
-$ doppler configs update --project proj --config dev --inheritable true
+$ doppler configs update --project proj --config dev --inheritable=true
 
 Configuring which configs the given config inherits
 Note: The inherits flag accepts a comma separated list of PROJ_NAME.CONF_NAME


```

File: pkg/http/api.go [hunk 3]
```diff
@@ -1050,6 +1050,90 @@ func UpdateConfig(host string, verifyTLS bool, apiKey string, project string, co
...
+	return info, Error{}
+}
+
+func UpdateConfigInherits(host string, verifyTLS bool, apiKey string, project string, config string, inherits string) (models.ConfigInfo, Error) {
+	inheritsObj := []models.ConfigDescriptor{}
+
+	if len(inherits) > 0 {
+		configDescriptors := strings.Split(inherits, ",")
+		for _, cd := range configDescriptors {
+			parts := strings.SplitN(cd, ".", 2)
```
Comment (nmanoogian) [inline pkg/http/api.go]: minor: If you remove the split limiter, we show the config descriptor error for more malformed strings. e.g. `database.dev.dev`

```diff
diff --git a/pkg/http/api.go b/pkg/http/api.go
index 92f7dd9..0414a58 100644
--- a/pkg/http/api.go
+++ b/pkg/http/api.go
@@ -1091,7 +1091,7 @@ func UpdateConfigInherits(host string, verifyTLS bool, apiKey string, project st
 	if len(inherits) > 0 {
 		configDescriptors := strings.Split(inherits, ",")
 		for _, cd := range configDescriptors {
-			parts := strings.SplitN(cd, ".", 2)
+			parts := strings.Split(cd, ".")
 			if len(parts) != 2 {
 				return models.ConfigInfo{}, Error{Message: "Config descriptors must match the format \"projectSlug.configName\""}
 			}


```

File: pkg/cmd/configs.go [hunk 4]
```diff
@@ -61,9 +61,19 @@ var configsDeleteCmd = &cobra.Command{
...
+
+Configuring which configs the given config inherits
+Note: The inherits flag accepts a comma separated list of PROJ_NAME.CONF_NAME
+$ doppler configs update --project proj --config dev --inherits "shared-db.dev,shared-api.dev"`,
```
Comment (nmanoogian) [inline pkg/cmd/configs.go]: There doesn't seem to be a way to clear the inherits list.

This was the first thing that came to mind for me to try but I'm not sure if this library can distinguish between this and an omitted flag:
```
$ doppler configs -p weather-service update -c dev --inherits ""
Error: flag needs an argument: --inherits
```

An alternative might be to have a sentinel value like `--inherits none`?

Comment (apazzolini) [reply pkg/cmd/configs.go]: To clear the inherits list, you can pass in `--inherits=`. I'll add some documentation.

**Review by nmanoogian — APPROVED** [review]

**Pull request merged**